src/ActionsLinux.cpp

```diff
@@ -0,0 +1,210 @@
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The source file %s could not be opened for reading: %s", src_file, strerror(errno));
+                break;
+            }
+
+            // Ensure that the target file is opened with a writable mode, even if the input file was readonly
+            out_file = open(target_file, O_WRONLY | O_CREAT | O_TRUNC, src_stat.st_mode | S_IWUSR);
+            if (out_file == -1)
+            {
+                result.m_ReturnCode = -1;
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The destination file %s could not be opened for writing: %s", src_file, strerror(errno));
+                break;
+            }
+
+            do {
+#ifdef FICLONE
+                // See if it is on the same filesystem (which it usually will be) - if it is, we can try using the FICLONE ioctl to do a lightweight copy-on-write copy
+                struct stat dst_stat;
+                if (fstat(out_file, &dst_stat) != 0)
+                {
+                    result.m_ReturnCode = -1;
+                    snprintf(tmpBuffer, sizeof(tmpBuffer), "The properties of the destination file %s could not be retrieved: %s", target_file, strerror(errno));
+                    break;
+                }
+
+                if (src_stat.st_dev == dst_stat.st_dev)
+                {
+                    // Try the IOCTL. We don't particularly care about errors here, we'll just fall back if this fails
+                    if (ioctl(out_file, FICLONE, in_file) != -1)
+                    {
+                        // It worked!
+                        break;
+                    }
+                }
+#endif
+
+                // Otherwise, next fastest method is to ask the kernel to do the copy using splice
+                size_t bytes_in, bytes_out;
+                do
+                {
+                    bytes_in = splice(in_file, NULL, temporary_pipe[1], NULL, src_stat.st_blksize, 0);
+                    if (bytes_in == -1)
+                    {
+                        result.m_ReturnCode = -1;
+                        snprintf(tmpBuffer, sizeof(tmpBuffer), "Reading from the source file %s failed: %s", src_file, strerror(errno));
+                        break;
+                    }
+
+                    bytes_out = splice(temporary_pipe[0], NULL, out_file, NULL, bytes_in, 0);
+                    if (bytes_out == -1)
+                    {
+                        result.m_ReturnCode = -1;
+                        snprintf(tmpBuffer, sizeof(tmpBuffer), "Writing to the destination file %s failed: %s", target_file, strerror(errno));
+                        break;
+                    }
+                } while (bytes_out > 0);
+
+            } while (false);
+
+            close(in_file);
+            in_file = -1;
+
+            close(out_file);
+            out_file = -1;
+
+            // Mark the stat cache dirty regardless of whether we failed or not - the target file is in an unknown state now
```

Comment:
No, if there were no errors then we _definitely_ need to mark the stat cache dirty, because we need the next attempt to stat the file to hit the disk and get the updated info. The case where we _might_ not need to mark it dirty is in the case of some failures - if we were able to open the output file but unable to splice any bytes to it, for example, then maybe it didn't actually change on disk - but it's better to assume the state is unknown.

Reply:
Better to start off safe and then optimise later if it's an issue.

Reply:
Yes. It's also not very important to optimize the case where a copy failed, I think.